File: 2016/Python Nov 2024/Day 3/day3.py
def main():
    fileName = "../TextFiles/day3.txt"
    partOne(fileName)
    partTwo(fileName)



def partOne(fileName):
    count = 0
    with open(fileName,"r") as f:
        for s in f.readlines():
            #print(s)
            (a ,b ,c) = s.split()
            if isValidTriangle(int(a),int(b),int(c)):
               count+=1
               #print((a,b,c),"is  a valid triangle")
            #else:
            #    print((a,b,c),"is not a valid triangle")
    print("Day 3 part 1:",count)

def isValidTriangle(a,b,c):
    if  a + b <= c:
        return False
    elif b + c <= a:
        return False
    elif c + a <= b:
        return False
    return True


def partTwo(fileName):
    count = 0
    with open(fileName,"r") as f:
        l1 = []
        l2 = []
        l3 = []

        for s in f.readlines():
            #print(s)
            (a ,b ,c) = s.split()
            l1.append(int(a))
            l2.append(int(b))
            l3.append(int(c))
        #l1.sort()
        #l2.sort()
        #l3.sort()
       # print(l1)




        for i in range(0,len(l1),3):
            #print(i)
            #print(l1[i],l1[i+1],l1[i+2])
            if isValidTriangle(l1[i],l1[i+1],l1[i+2]):
                count+= 1
        for i in range(0,len(l2),3):
            #print(l1[i],l1[i+1],l1[i+2])
            if isValidTriangle(l2[i],l2[i+1],l2[i+2]):
                count+= 1

        for i in range(0,len(l3),3):
            #print(l1[i],l1[i+1],l1[i+2])
            if isValidTriangle(l3[i],l3[i+1],l3[i+2]):
                count+= 1


    print("Day 3 part 2:",count)


main()

File: 2019/CPP Feb 2025/Day 3/day3.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <sstream>
#include <vector>

#include "day3.h"

int main()
{
    std::string fileName{ "../textFiles/day3final.txt" };
    partOne(fileName);
}

std::vector<std::vector<std::string>> getDirections(std::string fileName)
{
    std::ifstream file(fileName);
    std::vector<std::vector<std::string>> directions{};
    std::vector<std::string> direction{};
    std::string tok{};

    if (file.is_open())
    {
        std::string line;
        while (std::getline(file, line))
        {
            direction.push_back(line);
        }

        file.close();
        // std::cout << direction << std::endl;
        for (auto d : direction)
        {
            std::istringstream ss(d);
            std::vector<std::string> currentLine;
            while (std::getline(ss, tok, ','))
            {
                // std::cout << tok << std::endl;
                currentLine.push_back(tok);
            }
            directions.push_back(currentLine);
        }
    }
    else
    {
        std::cout << "unable to open file" << std::endl;
    }

    return directions;
}

void partOne(std::string fileName)
{
    std::vector<std::vector<std::string>> directions = getDirections(fileName);
    std::set<std::tuple<int, int>> firstCable{ getAllPoints(directions[0]) };
    std::set<std::tuple<int, int>> secondCable{ getAllPoints(directions[1]) };
    std::set<std::tuple<int, int>> intersectingPoint{};

    std::set_intersection(firstCable.begin(), firstCable.end(),
        secondCable.begin(), secondCable.end(),
        std::inserter(intersectingPoint, intersectingPoint.begin()));

    int minDist = INT_MAX;

    for (auto ip : intersectingPoint)
    {
        int newVal{ std::abs(std::get<0>(ip)) + std::abs(std::get<1>(ip)) };
        if (newVal < minDist && !(std::get<0>(ip) == 0 && std::get<1>(ip) == 0))
        {
            minDist = newVal;
        }
    }
    std::cout << "Day 3 part one: " << minDist << std::endl;
}

std::set<std::tuple<int, int>> getAllPoints(std::vector<std::string> directions)
{
    int startX = 0;
    int startY = 0;
    int steps{ 0 };
    std::set<std::tuple<int, int>> dictLeftRight{};

    for (std::string dir : directions)
    {
        steps = std::stoi(dir.substr(1));
        switch (dir[0])
        {
        case 'R':
            for (int i = 1; i <= steps; i++)
            {
                dictLeftRight.insert(std::tuple<int, int>(startX + i, startY));
            }
            startX += steps;
            break;
        case 'L':
            for (int i = 1; i <= steps; i++)
            {
                dictLeftRight.insert(std::tuple<int, int>(startX - i, startY));
            }
            startX -= steps;
            break;
        case 'U':

            for (int i = 1; i <= steps; i++)
            {
                dictLeftRight.insert(std::tuple<int, int>(startX, startY - i));
            }
            startY -= steps;
            break;
        case 'D':
            for (int i = 1; i <= steps; i++)
            {
                dictLeftRight.insert(std::tuple<int, int>(startX, startY + i));
            }
            startY += steps;
            break;
        }
    }
    return dictLeftRight;
}